### PR TITLE
[2D manifold in 3D] Option to use global Cartesian coordinates to define exact covariant basis vectors

### PR DESCRIPTION
**Pull request by tristanmontoya**
As mentioned in #49, there is a possible issue regarding the covariant basis vectors not being well defined at the poles when they are expressed in spherical coordinates. This is not the case when they are defined in Cartesian coordinates. The type parameter `GlobalCoordinateSystem` has now been added to `AbstractCovariantEquations` for dispatch purposes, in order to select which coordinate system should be used to construct the exact covariant basis vectors when initializing the auxiliary variables. This can be specified by the user in the constructor for the equations by passing in the keyword argument `global_coordinate_system`, which can be set to  `GlobalCartesianCoordinates()` (the default) or `GlobalSphericalCoordinates()`.

When a global Cartesian coordinate system is used, the `initial_condition_gaussian` struct now defines the solid body rotation as $\vec{v}(\vec{x}) = \vec{\omega} \times \vec{x}$, where $\vec{\omega}$ is the angular velocity vector, and computes the analytical solution using [Rodrigues' rotation formula](https://en.wikipedia.org/wiki/Rodrigues%27_rotation_formula), bypassing any use of spherical coordinates whatsoever.

**Note:** I personally do not see any practical downside to always using global Cartesian coordinates to define the covariant basis. It would simplify the code to just remove the option to use global spherical coordinates. However, the spherical approach appears to be more common in existing codes, and it may be valuable to have it as an option (and to facilitate comparisons of the approaches). If you have an opinion on whether to keep the spherical option or not, please comment on this PR. My current inclination is to keep it for now, and perhaps remove it later on if it becomes cumbersome to continue to support it.

### CODE REVIEW

**Comment by amrueda** [comment]
Great work!
I just tested convergence for modifications of the spherical advection elixir using the cubed sphere and icosahedral grids. The convergence was broken on the `main` branch for the icosahedral grid, probably due to https://github.com/trixi-framework/TrixiAtmo.jl/issues/49. The use of global Cartesian coordinates seems to fix this!

Here are the results I obtain with the `main` branch:
```julia
julia> convergence_test("../examples/elixir_spherical_advection_covariant.jl", 4, cells_per_dimension = (3,3))
####################################################################################################
l2
scalar              vcon1               vcon2               
error     EOC       error     EOC       error     EOC       
7.68e+00  -         0.00e+00  -         0.00e+00  -         
4.66e-01  4.04      0.00e+00  NaN       0.00e+00  NaN       
2.58e-02  4.18      0.00e+00  NaN       0.00e+00  NaN       
1.48e-03  4.12      0.00e+00  NaN       0.00e+00  NaN       

mean      4.11      mean      NaN       mean      NaN       
----------------------------------------------------------------------------------------------------
linf
scalar              vcon1               vcon2               
error     EOC       error     EOC       error     EOC       
6.14e+01  -         0.00e+00  -         0.00e+00  -         
6.55e+00  3.23      0.00e+00  NaN       0.00e+00  NaN       
7.56e-01  3.12      0.00e+00  NaN       0.00e+00  NaN       
4.93e-02  3.94      0.00e+00  NaN       0.00e+00  NaN       

mean      3.43      mean      NaN       mean      NaN       
----------------------------------------------------------------------------------------------------

julia> convergence_test("../examples/elixir_spherical_advection_covariant_icosahedron.jl", 4, cells_per_dimension = (1,1))
####################################################################################################
l2
scalar              vcon1               vcon2               
error     EOC       error     EOC       error     EOC       
8.74e+00  -         0.00e+00  -         0.00e+00  -         
2.46e+00  1.83      0.00e+00  NaN       0.00e+00  NaN       
1.21e+00  1.02      0.00e+00  NaN       0.00e+00  NaN       
6.53e-01  0.89      0.00e+00  NaN       0.00e+00  NaN       

mean      1.25      mean      NaN       mean      NaN       
----------------------------------------------------------------------------------------------------
linf
scalar              vcon1               vcon2               
error     EOC       error     EOC       error     EOC       
8.08e+01  -         0.00e+00  -         0.00e+00  -         
3.84e+01  1.07      0.00e+00  NaN       0.00e+00  NaN       
2.43e+01  0.66      0.00e+00  NaN       0.00e+00  NaN       
1.62e+01  0.58      0.00e+00  NaN       0.00e+00  NaN       

mean      0.77      mean      NaN       mean      NaN       
----------------------------------------------------------------------------------------------------
Dict{Symbol, Any} with 3 entries:
  :variables => ("scalar", "vcon1", "vcon2")
  :l2        => [1.24716, NaN, NaN]
  :linf      => [0.771628, NaN, NaN]
```

And here are the results I obtain with `tm/new_covariant_metrics`:
```julia
julia> convergence_test("../examples/elixir_spherical_advection_covariant.jl", 4, cells_per_dimension = (3,3))
####################################################################################################
l2
scalar              vcon1               vcon2               vcon3               
error     EOC       error     EOC       error     EOC       error     EOC       
7.68e+00  -         0.00e+00  -         0.00e+00  -         0.00e+00  -         
4.66e-01  4.04      0.00e+00  NaN       0.00e+00  NaN       0.00e+00  NaN       
2.58e-02  4.18      0.00e+00  NaN       0.00e+00  NaN       0.00e+00  NaN       
1.48e-03  4.12      0.00e+00  NaN       0.00e+00  NaN       0.00e+00  NaN       

mean      4.11      mean      NaN       mean      NaN       mean      NaN       
----------------------------------------------------------------------------------------------------
linf
scalar              vcon1               vcon2               vcon3               
error     EOC       error     EOC       error     EOC       error     EOC       
6.14e+01  -         0.00e+00  -         0.00e+00  -         0.00e+00  -         
6.55e+00  3.23      0.00e+00  NaN       0.00e+00  NaN       0.00e+00  NaN       
7.56e-01  3.12      0.00e+00  NaN       0.00e+00  NaN       0.00e+00  NaN       
4.93e-02  3.94      0.00e+00  NaN       0.00e+00  NaN       0.00e+00  NaN       

mean      3.43      mean      NaN       mean      NaN       mean      NaN       
----------------------------------------------------------------------------------------------------
Dict{Symbol, Any} with 3 entries:
  :variables => ("scalar", "vcon1", "vcon2", "vcon3")
  :l2        => [4.11262, NaN, NaN, NaN]
  :linf      => [3.42795, NaN, NaN, NaN]

julia> convergence_test("../examples/elixir_spherical_advection_covariant_icosahedron_newbranch.jl", 4, cells_per_dimension = (1,1))
####################################################################################################
l2
scalar              vcon1               vcon2               vcon3               
error     EOC       error     EOC       error     EOC       error     EOC       
7.49e+00  -         0.00e+00  -         0.00e+00  -         0.00e+00  -         
5.18e-01  3.85      0.00e+00  NaN       0.00e+00  NaN       0.00e+00  NaN       
4.76e-02  3.45      0.00e+00  NaN       0.00e+00  NaN       0.00e+00  NaN       
3.67e-03  3.69      0.00e+00  NaN       0.00e+00  NaN       0.00e+00  NaN       

mean      3.66      mean      NaN       mean      NaN       mean      NaN       
----------------------------------------------------------------------------------------------------
linf
scalar              vcon1               vcon2               vcon3               
error     EOC       error     EOC       error     EOC       error     EOC       
6.64e+01  -         0.00e+00  -         0.00e+00  -         0.00e+00  -         
1.35e+01  2.29      0.00e+00  NaN       0.00e+00  NaN       0.00e+00  NaN       
1.67e+00  3.02      0.00e+00  NaN       0.00e+00  NaN       0.00e+00  NaN       
1.33e-01  3.66      0.00e+00  NaN       0.00e+00  NaN       0.00e+00  NaN       

mean      2.99      mean      NaN       mean      NaN       mean      NaN       
----------------------------------------------------------------------------------------------------
Dict{Symbol, Any} with 3 entries:
  :variables => ("scalar", "vcon1", "vcon2", "vcon3")
  :l2        => [3.66488, NaN, NaN, NaN]
  :linf      => [2.98918, NaN, NaN, NaN]
```

File: src/equations/covariant_advection.jl [hunk 1]
```diff
@@ -14,75 +15,102 @@ as a system of equations in which the first variable is the scalar conserved qua
...
+    function CovariantLinearAdvectionEquation2D(;
+                                                global_coordinate_system = GlobalCartesianCoordinates())
+        return new{typeof(global_coordinate_system)}(global_coordinate_system)
+    end
+end
 
+# The conservative variables are the scalar conserved quantity and three contravariant 
+# velocity components. Note that vcon3 is zero, since the velocity is always tangent to the 
+# surface, but we keep it because there may be 3 global velocity components (if global 
+# Cartesian coordinates are used), and the number of variables needs to match.
```
Comment (amrueda) [inline src/equations/covariant_advection.jl:60]: Please adjust if we remove vcon3

File: src/equations/covariant_advection.jl [hunk 2]
```diff
@@ -14,75 +15,102 @@ as a system of equations in which the first variable is the scalar conserved qua
...
-are spatially varying but assumed to be constant in time, so we do not apply any flux or 
-dissipation to such variables. The resulting system is then given on the reference element 
-as 
+and $\xi^1$ and $\xi^2$ are the local reference space coordinates. The third velocity 
+component $v^3$, representing the velocity in the direction normal to the surface, is 
+stored as a fourth variable which is set to zero when solving equations on two-dimensional 
+surfaces in three-dimensional ambient space, and will remain zero throughout the simulation 
+(we use three velocity components to match the output of [`contravariant2global`](@ref), 
+which returns a three-dimensional velocity vector in the global Cartesian or spherical 
+coordinate system). The velocity components are spatially varying but assumed to be
```
[the remark this answers is not in the document]
Comment (tristanmontoya) [reply src/equations/covariant_advection.jl]: I have removed `vcon3` but have kept three global velocity components; I didn't want the number of fields to depend on the choice of global coordinate system, and in the future we will have a vertical velocity that we'll need to output anyway.

File: src/equations/equations.jl [hunk 3]
```diff
@@ -48,48 +69,122 @@ dispatching on the return type.
...
+of local contravariant components into the global coordinate system specified by the 
+`GlobalCoordinateSystem` type parameter in [`AbstractCovariantEquations`](@ref). `u` is a 
+vector type of the correct length `nvariables(equations)`. Notice the function doesn't 
+include any error checks for the purpose of efficiency, so please make sure your input is 
+correct. The inverse conversion is performed by [`global2contravariant`](@ref).
+"""
+function contravariant2global end
+
+"""
+    contravariant2global(u, aux_vars, equations)
```
[the remark this answers is not in the document]
Comment (tristanmontoya) [reply src/equations/equations.jl]: Fixed in [a50c90d](https://github.com/trixi-framework/TrixiAtmo.jl/pull/54/commits/a50c90de0827ae97d303c584d1e6c63ab609815c).

File: src/equations/covariant_advection.jl [hunk 4]
```diff
@@ -47,42 +60,44 @@ function velocity(u, ::CovariantLinearAdvectionEquation2D)
...
+# slot, with zeros in all other positions
 @inline function Trixi.cons2entropy(u, aux_vars,
                                     equations::CovariantLinearAdvectionEquation2D)
     z = zero(eltype(u))
-    return SVector{3}(u[1], z, z)
+    return SVector(u[1], z, z)
 end
 
 # Numerical flux plus dissipation for abstract covariant equations as a function of the 
 # state vector u, as well as the auxiliary variables aux_vars, which contain the geometric
```
Comment (amrueda) [inline src/equations/covariant_advection.jl:93]:
```suggestion
# Flux for abstract covariant equations as a function of the 
# state vector u, as well as the auxiliary variables aux_vars, which contain the geometric 
```

File: src/equations/equations.jl [hunk 5]
```diff
@@ -48,48 +69,122 @@ dispatching on the return type.
...
+`nvariables(equations)`. Notice the function doesn't include any error checks for the 
+purpose of efficiency, so please make sure your input is correct. The inverse conversion is 
+performed by [`contravariant2global`](@ref).
+"""
+function global2contravariant end
 
 # If no auxiliary variables are passed into the conversion to spherical coordinates, do not 
 # do any conversion.
-@inline contravariant2spherical(u, equations) = u
+@inline contravariant2global(u, ::AbstractEquations) = u
```
Comment (amrueda) [inline src/equations/equations.jl:137]: Not needed anymore?
```suggestion
```

**Review by amrueda — APPROVED** [review]
LGTM!

File: src/equations/equations.jl (hunk 5 above)
Comment (tristanmontoya) [reply src/equations/equations.jl:137]: Yes, this is not needed. I will delete and then merge.

**Pull request merged**